lib/AST/ASTPrinter.cpp

```diff
@@ -5156,6 +5203,103 @@ void PrintAST::visitErasureExpr(ErasureExpr *expr) {
 }
 
 void PrintAST::visitKeyPathExpr(KeyPathExpr *expr) {
+  if (expr->isObjC()) {
+    Printer << "#keyPath(";
+  }
+  
+  if (!expr->isObjC()) {
+    if (auto rootType = expr->getRootType()) {
+      Printer << "\\";
+      printTypeLoc(rootType);
+    } else {
+      visit(expr->getParsedRoot());
+    }
+  }
+  
+  auto components = expr->getComponents();
+  if (!components.empty()) {
+    for (auto& component: components) {
+      auto kind = component.getKind();
+      switch (kind) {
+        case swift::KeyPathExpr::Component::Kind::Invalid: {
+          break;
+        }
+        case swift::KeyPathExpr::Component::Kind::UnresolvedProperty: {
+          Printer << "." << component.getUnresolvedDeclName();
+          break;
+        }
+        case swift::KeyPathExpr::Component::Kind::UnresolvedSubscript: {
+          auto args = component.getSubscriptArgs();
+          Printer << ".";
+          
+          
+          Printer << "[";
+          // component.forUnresolvedSubscript(<#ASTContext &ctx#>, args)
+          Printer << "]";
```

Comment:
You'll want to print the arguments here, similar to printing for SubscriptExpr. In fact it may be worth factoring out an ArgumentList printing function at this point that you could share between CallExpr, SubscriptExpr, and here. You could have it take a `bool forSubscript` argument to change printing from `(...)` to `[...]`.